### PR TITLE
[WIP] Subgraph op

### PR DESCRIPTION
**Pull request by reminisce**
## Description ##
@zheng-da 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/subgraph/partition_graph.cc [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+  }
+  SimpleNode() : label(-1), node(nullptr) {}
+  int label;
+  nnvm::Node* node;
+  std::unordered_map<Node*, int> outputs;
+  //std::unordered_map<SimpleNodePtr, int> inputs;
+  //std::unordered_map<SimpleNodePtr, int> outputs;
+};
+
+void CreateSimpleGraph(const Graph& g,
```
Comment (reminisce) [inline src/operator/subgraph/partition_graph.cc]: @zheng-da Call this to create an undirected graph from `g`.

File: src/operator/subgraph/partition_graph.cc [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        } else {
+          CHECK_EQ(simple_nodes[nid]->label, label);
+        }
+      }
+    }
+  }
+}
+
+// number of subgraphs found
+void FindSubgraphs(const Graph& g,
```
Comment (reminisce) [inline src/operator/subgraph/partition_graph.cc]: @zheng-da Call this to find all subgraphs.

File: src/operator/subgraph/partition_graph.cc [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    nnvm::Node* node = simple_nodes[i]->node;
+    if (!node->is_variable() && simple_nodes[i]->label == -1 && op_names.count(node->op()->name)) {
+      subgraph_nodes->emplace_back();
+      LabelSubgraph(g, op_names, subgraph_nodes->size() - 1, i, simple_nodes, &subgraph_nodes->back());
+    }
+  }
+}
+
+// find the input entries of a subgraph
+void FindInputEntries(const Graph& g,
```
Comment (reminisce) [inline src/operator/subgraph/partition_graph.cc]: @zheng-da

File: src/operator/subgraph/partition_graph.cc [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+        input_entries->push_back(&e);
+      } else {
+        CHECK_EQ(simple_nodes[nid]->label, label);
+      }
+    }
+  }
+}
+
+// find the output entries of a subgraph
+void FindOutputEntries(const Graph& g,
```
Comment (reminisce) [inline src/operator/subgraph/partition_graph.cc]: @zheng-da

File: src/operator/subgraph/partition_graph.cc [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ *  Copyright (c) 2016 by Contributors
+ * \file quantization.cc
```
Comment (zheng-da) [inline src/operator/subgraph/partition_graph.cc]: file name is incorrect.

File: src/operator/subgraph/partition_graph.cc [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+  //std::unordered_map<SimpleNodePtr, int> outputs;
+};
+
+void CreateSimpleGraph(const Graph& g,
+                       std::vector<SimpleNodePtr>* simple_nodes) {
+  const auto& indexed_graph = g.indexed_graph();
+  simple_nodes->reserve(indexed_graph.num_nodes());
+  for (size_t nid = 0; nid < indexed_graph.num_nodes(); ++nid) {
+    SimpleNodePtr sn = SimpleNode::Create();
+    sn->node = const_cast<nnvm::Node*>(indexed_graph[nid].source);
```
Comment (zheng-da) [inline src/operator/subgraph/partition_graph.cc]: this is kind of hacky. i think it's better to call DFSVisit to get nodes.

File: src/operator/subgraph/partition_graph.cc [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+  });
+#endif
+}
+
+void LabelSubgraph(const Graph&g,
+                   const std::unordered_set<std::string>& op_names,
+                   const int label,
+                   const size_t snid,  // simple node id
+                   const std::vector<SimpleNodePtr>& simple_nodes,
+                   std::vector<SimpleNode*>* subgraph_nodes) {
```
Comment (zheng-da) [inline src/operator/subgraph/partition_graph.cc]: can you add comments on this function?
my understanding is that this function labels nodes that connect to a starting node via both input links and output links.

Comment (zheng-da) [reply src/operator/subgraph/partition_graph.cc]: You could generalize this function by
* generalize the node selection criteria (e.g., passing a functor),
* allow users to specify following input links, output links or both.

File: src/operator/subgraph/partition_graph.cc [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    subgraph_node_ptr->inputs.emplace_back(new_node_entry);
+  }
+  Graph ret;
+  static const auto& flist_outputs = nnvm::Op::GetAttr<nnvm::FListOutputNames>("FListOutputNames");
+  auto list_output_names_func = flist_outputs.get(subgraph_node_ptr->op(), nullptr);
+  CHECK(list_output_names_func != nullptr);
+  const size_t num_outputs = list_output_names_func(subgraph_node_ptr->attrs).size();
+  for (uint32_t i = 0; i < num_outputs; ++i) {
+    ret.outputs.emplace_back(nnvm::NodeEntry{subgraph_node_ptr, i, 0});
+  }
```
Comment (zheng-da) [inline src/operator/subgraph/partition_graph.cc]: i suppose this function isn't complete yet?

File: src/operator/subgraph/partition_graph.cc (hunk 6 above)
Comment (reminisce) [reply src/operator/subgraph/partition_graph.cc]: Yes, this is a quick implementation for getting the flow to run through. Will get rid of this later.

File: src/operator/subgraph/partition_graph.cc (hunk 7 above)
Comment (reminisce) [reply src/operator/subgraph/partition_graph.cc]: Agree. For now I only focus on getting the flow to run through in my use case. After sorting out every detail of the implementation, we can revisit and come up with a generic interface design.

**Comment by reminisce** [comment]
Graph partitioning example of replacing operators: Convolution, BatchNorm, Pooling, and Relu with subgraph operators in Resnet-152.
- Original graph
image

- Partitioned graph
image
